stages(kickstart): implement "reboot" option

mvo5 opened the pull request:
This commit implements the `reboot` option for kickstart files.

Note that there are two ways this can be enabled via the json.
Either via a boolean or by passing a dict with options.
```
{"reboot": true}
{"reboot": {"eject": true, "kexec": true}
```

Passing the empty dict
```
{"reboot": {}}
```
is not allowed by the schema.

supakeen reviewed (commented):
There's a typo in the commit message:

> "clearpart" have have the same behavior

Other than that I think we can do something like requiring at least one property to be passed when `eject` is an object to alleviate the thinking about 'what does an empty dict mean'.

mvo5 commented:
> There's a typo in the commit message:
> 
> > "clearpart" have have the same behavior
> 
> Other than that I think we can do something like requiring at least one property to be passed when `eject` is an object to alleviate the thinking about 'what does an empty dict mean'.

Great suggestion! I changed the code (and schema) this way now that the empty dict is not allowed.

achilleas-k reviewed (approved):
Nice.  LGTM!

achilleas-k commented:
The `clearpart` test is failing.  It seems it's not technically wrong to have `clearpart` without an option:
```
$ cat test.ks
clearpart
$ ksvalidator test.ks

Checking kickstart file test.ks

$
```

achilleas-k commented:
> requiring at least one property to be passed when `eject` is an object

I was a bit confused here.  Did you mean `reboot`?

mvo5 commented:
> The `clearpart` test is failing. It seems it's not technically wrong to have `clearpart` without an option:
> 
> ```
> $ cat test.ks
> clearpart
> $ ksvalidator test.ks
> 
> Checking kickstart file test.ks
> 
> $
> ```

Sorry! That was my bad, I pushed this a bit too hastily (right before a meeting, I really should have learned by now that doing anything in a rush is a bad idea).  

I fixed it now to make an empty `clearpart` an invalid option. Technically it's fine to have it without an option but the kickstart docs say: `Removes partitions from the system, prior to creation of new partitions. By default, no partitions are removed.`  which I read as "it won't do anything without an option" and requiring an option avoid the awkward question if an empty dict is valid or not.


> There's a typo in the commit message:
> 
> > "clearpart" have have the same behavior
> 
> Other than that I think we can do something like requiring at least one property to be passed when `eject` is an object to alleviate the thinking about 'what does an empty dict mean'.

Thanks, I fixed the typo (by removing this part of the commit message and split the commit in two as it should have been in the first place, sorry for that). Also thanks for the suggestion about requiring that a dict for `reboot` needs to have at least one option, I really like this !

teg commented:
Yay for more ks options!

For my own education: is this needed by one of our images, or is this done for completeness?

achilleas-k commented:
> For my own education: is this needed by one of our images, or is this done for completeness?

It's needed to replicate the Edge ISO kickstart.
https://github.com/RedHatInsights/edge-api/blob/cab94ee6ed753c7bfe5b6ecfba314ff6a6cb7f59/templates/templateKickstart.ks#L1-L9
